pkg/registry/pod/strategy.go

```diff
@@ -65,6 +65,21 @@ func (podStrategy) PrepareForUpdate(obj, old runtime.Object) {
 	newPod := obj.(*api.Pod)
 	oldPod := old.(*api.Pod)
 	newPod.Status = oldPod.Status
+	// TODO: VolumeMounts of init containers currently aren't merged properly
+	// by the PATCH handler because they are stored as a string under
+	// PodInitContainersAnnotationKey (pod.alpha.kubernetes.io/init-containers)
+	// in annotations of v1.PodSpec.
+	// In the existing Pod, VolumeMounts of init containers may be altered by
+	// secrets / config maps mechanisms causing ValidatePodUpdate to fail.
+	// As a workaround, changes in VolumeMounts of init containers are
+	// currently ignored and don't cause validation errors.
+	// Remove this workaround when InitContainers go out of alpha and
+	// PodInitContainersAnnotationKey is removed.
+	if len(oldPod.Spec.InitContainers) == len(newPod.Spec.InitContainers) {
```

Comment:
@liggitt that'a side effect of init containers being represented as an annotation key. Because of this, VolumeMounts property isn't merged properly and the apiserver thinks that you're trying to change VolumeMounts even when you don't. On the server side, you may have e.g. secrets added to VolumeMounts. Then you submit a patch for a pod that say changes the image of an init container (which you can), but because merge mechanisms don't work for the annotation validation code thinks that you're also trying to change VolumeMounts by replacing them with an empty list that was deserialized from the annotation. That's temporary thing, I've added references to `PodInitContainersAnnotationKey` and `pod.alpha.kubernetes.io/init-containers` to the comment so when these annotations are removed this part also shows up during grepping and this workaround can be removed, too.